### PR TITLE
Use `config[]` not `config.get()`

### PR DESCRIPTION
**Pull request by willbarrett**
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [X] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

As pointed out here https://github.com/apache/incubator-superset/pull/8409/files#diff-6a3371457528722a734f3c51d9238c13R482 we no longer want to use the `config.get` format, but instead assume that all configuration options are set. This PR alters all existing usages of `config.get` to the `config[]` syntax. Note that this will cause errors if those configuration options are not provided. In order to get the system running again, I had to set the following configuration options locally:

```python
# Stuff that needs defaulting
ENABLE_CHUNK_ENCODING = False
FEATURE_FLAGS = {}
EVENT_LOGGER = None
SQLALCHEMY_CUSTOM_PASSWORD_STORE = None
QUERY_LOGGER = None
FAB_ADD_SECURITY_VIEWS = True
SUPERSET_LOG_VIEW = True
HTTP_HEADERS = {}
```

I'm somewhat concerned about the effect of merging this in terms of stability, so I'm looking for feedback from the community, especially @villebro who clued me in to the preferred approach in the first place.

### TEST PLAN

- Run CI 
- Click around locally
- Request feedback from the community

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@villebro @mistercrunch @dpgaspar

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -109,7 +109,7 @@ For large features or major changes to codebase, please create **Superset Improv
 
 ### Fix Bugs
 
-Look through the GitHub issues. Issues tagged with `#bug` is
+Look through the GitHub issues. Issues tagged with `#bug` are
 open to whoever wants to implement it.
```
Comment (villebro) [inline CONTRIBUTING.md]: ..wants to implement them.

File: superset/__init__.py [hunk 2]
```diff
@@ -195,21 +195,21 @@ def index(self):
...
 
-results_backend = app.config.get("RESULTS_BACKEND")
-results_backend_use_msgpack = app.config.get("RESULTS_BACKEND_USE_MSGPACK")
+results_backend = app.config["RESULTS_BACKEND"]
+results_backend_use_msgpack = app.config["RESULTS_BACKEND_USE_MSGPACK"]
 
 # Merge user defined feature flags with default feature flags
-_feature_flags = app.config.get("DEFAULT_FEATURE_FLAGS") or {}
-_feature_flags.update(app.config.get("FEATURE_FLAGS") or {})
+_feature_flags = app.config["DEFAULT_FEATURE_FLAGS"] or {}
```
Comment (villebro) [inline superset/__init__.py]: I don't think we need the `or {}` here, as `DEFAULT_FEATURE_FLAGS` is `Dict[str, bool]` in `config.py`.

File: superset/__init__.py [hunk 3]
```diff
@@ -195,21 +195,21 @@ def index(self):
...
+_feature_flags.update(app.config["FEATURE_FLAGS"])
 
 # Event Logger
 event_logger = get_event_logger_from_cfg_value(
-    app.config.get("EVENT_LOGGER", DBEventLogger())
+    app.config["EVENT_LOGGER"] or DBEventLogger()
```
Comment (villebro) [inline superset/__init__.py]: I believe the default value here should be defined in `config.py`

File: superset/common/query_object.py [hunk 4]
```diff
@@ -66,8 +66,8 @@ def __init__(
         extras: Optional[Dict] = None,
         columns: Optional[List[str]] = None,
         orderby: Optional[List[List]] = None,
-        relative_start: str = app.config.get("DEFAULT_RELATIVE_START_TIME", "today"),
-        relative_end: str = app.config.get("DEFAULT_RELATIVE_END_TIME", "today"),
+        relative_start: str = app.config["DEFAULT_RELATIVE_START_TIME"],
+        relative_end: str = app.config["DEFAULT_RELATIVE_END_TIME"],
```
Comment (villebro) [inline superset/common/query_object.py]: This is a good example of why this change is good; no point in having secondary default values spread across the codebase, as these should only be defaulted in one place.

File: superset/sql_lab.py [hunk 5]
```diff
@@ -50,10 +50,10 @@
 from superset.utils.decorators import stats_timing
 
 config = app.config
-stats_logger = config.get("STATS_LOGGER")
-SQLLAB_TIMEOUT = config.get("SQLLAB_ASYNC_TIME_LIMIT_SEC", 600)
+stats_logger = config["STATS_LOGGER"]
+SQLLAB_TIMEOUT = config["SQLLAB_ASYNC_TIME_LIMIT_SEC"]
 SQLLAB_HARD_TIMEOUT = SQLLAB_TIMEOUT + 60
```
Comment (villebro) [inline superset/sql_lab.py]: Out of scope, but I wonder if `SQLLAB_HARD_TIMEOUT` should be defined in `config.py`, too?

File: superset/sql_lab.py [hunk 6]
```diff
@@ -400,7 +400,7 @@ def execute_sql_statements(
                 )
             cache_timeout = database.cache_timeout
             if cache_timeout is None:
-                cache_timeout = config.get("CACHE_DEFAULT_TIMEOUT", 0)
+                cache_timeout = config["CACHE_DEFAULT_TIMEOUT"]
```
Comment (villebro) [inline superset/sql_lab.py]: I wonder what the point of defaulting to 0 was here. I can only assume it is a relic from long ago.

File: superset/config.py [hunk 7]
```diff
@@ -522,7 +522,7 @@ class CeleryConfig(object):
 SMTP_MAIL_FROM = "superset@superset.com"
 
 if not CACHE_DEFAULT_TIMEOUT:
-    CACHE_DEFAULT_TIMEOUT = CACHE_CONFIG.get("CACHE_DEFAULT_TIMEOUT")  # type: ignore
+    CACHE_DEFAULT_TIMEOUT = CACHE_CONFIG["CACHE_DEFAULT_TIMEOUT"]  # type: ignore
```
Comment (villebro) [inline superset/config.py]: The parameters you mentioned in the description that needed to be defaulted (`ENABLE_CHUNK_ENCODING` etc) seem to be missing here.

**Review by villebro — CHANGES_REQUESTED** [review]
@john-bodley can probably chime in here, too, but I think in general there are lots of good changes here, i.e. removing default settings that are scattered across the codebase that should be defined only in `config.py`. Also the fact that you were able to find eight (!) missing config parameters by doing this change goes to show that this change makes the codebase more robust. Some small changes proposed, but apart from that LGTM.

File: superset/config.py (hunk 7 above)
Comment (john-bodley) [reply superset/config.py]: Do we still need the `# type: ignore`?

**Review by john-bodley — COMMENTED** [review]
Thanks for tackling this. You mentioned there were a number of variables you had to set locally. Any reason they haven’t been added to `config.py`?

File: superset/sql_lab.py (hunk 5 above)
Comment (etr2460) [reply superset/sql_lab.py]: I don't think sql lab hard timeout needs to be definable in the config as well, the purpose for it is to provide a hard cutoff in celery if a soft timeout fails (or isn't implemented). The time between soft and hard timeout shouldn't be much, so I don't think allowing configuration gets much here. Happy to reconsider if you think there's a good reason to make it configurable though!

**Comment by etr2460** [comment]
+1 to @john-bodley's point, if switching from `get` to `[]` causes errors because the variables weren't defined in the config, then I think we should define defaults in the `config.py` file.

Also, in terms of typing, once we move to python 3.8, I think defining a Config type as a `TypedDict` (https://www.python.org/dev/peps/pep-0589/) should be really helpful here, as it'll catch any config vars not set.

I think this is an awesome change overall though, thanks for taking it on!

File: superset/sql_lab.py (hunk 6 above)
Comment (willbarrett) [reply superset/sql_lab.py]: It must be a relic, as it is set in config.py.

**Comment by willbarrett** [comment]
Thank you @villebro, @john-bodley, and @etr2460 for the feedback. The only reason the additional configuration wasn't added to `config.py` was that I'm new to the repo and wasn't aware of the correct place to put it :). Your guidance was helpful. You should find the missing configuration options added to `config.py`.

**Comment by mistercrunch** [comment]
It'd be nice if we could raise a `PresetConfigError` instead of a `KeyError` but that may be tricky since Flask manages the `config` class.

**Comment by willbarrett** [comment]
@mistercrunch yes, throwing a custom configuration error would be a nice improvement. I'd like to attempt that in a separate PR once this guy goes through - it may take a little doing to separate out Superset config from Flask config or extend Flask's config object to provide that functionality.

**Comment by willbarrett** [comment]
Hey there @villebro @etr2460 @john-bodley - I think this is ready for another look when you have time. Thanks!

File: superset/sql_lab.py (hunk 5 above)
Comment (villebro) [reply superset/sql_lab.py]: Not feeling especially opinionated on this one, but in general I have an aversion to hard coding parameters, especially in the context of distributed/async problems which are extra horrible to debug. So, at the risk of cluttering the config file, I would still vote for introducing the config parameter `SQLLAB_HARD_TIMEOUT_ADDON` or similar with a concise comment might save someone down the line a few hours of grief.

**But**: this ended up sounding much more dramatic than it probably is, so happy either way!

File: superset/views/core.py [hunk 8]
```diff
@@ -1491,7 +1487,7 @@ def get_datasource_label(ds_name: utils.DatasourceName) -> str:
             tables = [tn for tn in tables if tn.schema in valid_schemas]
             views = [vn for vn in views if vn.schema in valid_schemas]
 
-        max_items = config.get("MAX_TABLE_NAMES") or len(tables)
+        max_items = config["MAX_TABLE_NAMES"] or len(tables)
```
Comment (villebro) [inline superset/views/core.py]: nit: isn't `..or tables` sufficient here?

File: tests/celery_tests.py [hunk 9]
```diff
@@ -32,12 +32,12 @@
...
 from .base_tests import SupersetTestCase
 
-BASE_DIR = app.config.get("BASE_DIR")
+BASE_DIR = app.config["BASE_DIR"]
 CELERY_SLEEP_TIME = 5
 
 
 class CeleryConfig(object):
-    BROKER_URL = app.config.get("CELERY_RESULT_BACKEND")
+    BROKER_URL = app.config["CELERY_CONFIG"].BROKER_URL
```
Comment (villebro) [inline tests/celery_tests.py]: Yet another great catch by this change!

**Review by villebro — APPROVED** [review]
LGTM apart from a few non-blocking comments.

File: superset/views/core.py (hunk 8 above)
Comment (willbarrett) [reply superset/views/core.py]: I think it's required because `max_items` is used as an index in a slice later on. I may be mistaken, and would love an education if so!

Comment (villebro) [reply superset/views/core.py]: Sorry, my bad!

**Review by etr2460 — APPROVED** [review]

**Comment by willbarrett** [comment]
Hi @john-bodley - I think you have the remaining review here. Please forgive the reminder :)

**Review by john-bodley — APPROVED** [review]
